common/src/androidTest/java/com/microsoft/identity/common/MsalCppOAuth2TokenCacheTest.java

```diff
@@ -0,0 +1,249 @@
+import static com.microsoft.identity.common.SharedPreferencesAccountCredentialCacheTest.TARGET;
+import static com.microsoft.identity.common.SharedPreferencesAccountCredentialCacheTest.USERNAME;
+
+@RunWith(AndroidJUnit4.class)
+public class MsalCppOAuth2TokenCacheTest extends AndroidSecretKeyEnabledHelper {
+
+    private MsalCppOAuth2TokenCache mCppCache;
+
+    // Test Accounts/Credentials
+    private AccountCredentialTestBundle mTestBundle;
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+
+        // Context and related init
+        final Context context = InstrumentationRegistry.getTargetContext();
+        mCppCache = MsalCppOAuth2TokenCache.create(context);
+
+        // Credentials for testing
+        mTestBundle = new AccountCredentialTestBundle(
+                MicrosoftAccount.AUTHORITY_TYPE_V1_V2,
+                LOCAL_ACCOUNT_ID,
+                USERNAME,
+                HOME_ACCOUNT_ID,
+                ENVIRONMENT,
+                REALM,
+                TARGET,
+                CACHED_AT,
+                EXPIRES_ON,
+                SECRET,
+                CLIENT_ID,
+                SECRET,
+                MOCK_ID_TOKEN_WITH_CLAIMS,
+                null,
+                CredentialType.V1IdToken
+        );
+    }
+
+    @After
+    public void tearDown() {
+        mCppCache.clearCache();
+    }
+
+    @Test
+    public void saveAndGetAccountTest() throws ClientException {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+
+        // Save the the cache
+        mCppCache.saveAccountRecord(generatedAccount);
+
+        // Restore it
+        final AccountRecord restoredAccount = mCppCache.getAccount(
+                generatedAccount.getHomeAccountId(),
+                generatedAccount.getEnvironment(),
+                generatedAccount.getRealm()
+        );
+
+        Assert.assertNotNull(restoredAccount);
+        Assert.assertEquals(generatedAccount, restoredAccount);
+    }
+
+    @Test
+    public void getAccountNullTest() throws ClientException {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+        final AccountRecord restoredAccount = mCppCache.getAccount(
+                generatedAccount.getHomeAccountId(),
+                generatedAccount.getEnvironment(),
+                generatedAccount.getRealm()
+        );
+        Assert.assertNull(restoredAccount);
+    }
+
+    @Test
+    public void getAllAccountsTest() {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+        mCppCache.saveAccountRecord(generatedAccount);
+
+        final List<AccountRecord> accounts = mCppCache.getAllAccounts();
+        final AccountRecord restoredAccount = accounts.get(0);
+        Assert.assertEquals(generatedAccount, restoredAccount);
+
+        // TODO this API makes no statement about the immutability of the List returned
```

Comment:
👀

Reply:
What happens if you write to it?  Should we have standard tests against our interface packages that all implementations must pass?

Reply:
It looks like it delegates here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/c2dedc4786862ea48f67bf00f85c0c1e3d9ec127/common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java#L1039 which is an unmodifiable wrapper.

The other implementation is SharedPreferencesAccountCredentialCache, which returns a safe-copy.  https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/c2dedc4786862ea48f67bf00f85c0c1e3d9ec127/common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java#L223 We should pick one strategy.  And document it. 

And, in relation to my other comment, do such things *need* to return a list?

Reply:
I am leaning towards a standardized approach to the APIs whereby we designate the results are read-only and enforced via some unmodifiable wrapper is my thinking. There's a bit of blend of these in the code, as you're noticing.

IMO we should document on the API (via javadoc) and encourage our developers/contributors to do the same; we should be diligent about it in our code reviews. W.r.t to where else we could/should document such style decisions, the contributors guide/wiki seem like good candidates.

Reply:
Yes.  I'm a huge proponent of the ImmutableObjects paradigm, if it hasn't become apparent.  It simplifies a lot of thread-safety concerns, though it does imply behaviour changes on the part of clients.  We probably can't get there from here without a MV release, and it might be moot via OneAuth and MSAL-cpp.